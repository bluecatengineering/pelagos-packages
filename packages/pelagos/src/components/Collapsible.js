import {cloneElement, useMemo} from 'react';
import PropTypes from 'prop-types';
import ChevronDown from '@carbon/icons-react/es/ChevronDown';

import addClassName from '../functions/addClassName';
import cloneWithClassName from '../functions/cloneWithClassName';
import useCollapse from '../hooks/useCollapse';

import './Collapsible.less';

/** A collapsible container. */
const Collapsible = ({id, className, open, children, onHeaderClick}) => {
	id = useMemo(() => id || 'e' + ('' + Math.random()).slice(2), [id]);
	const headerId = id + '-header';
	const contentId = id + '-content';
	const headerElement = cloneWithClassName(children[0], 'Collapsible__title');
	const contentElement = children[1];
	return (
		<div id={id} className={`Collapsible${className ? ` ${className}` : ''}`}>
			<button
				id={headerId}
				className="Collapsible__header"
				type="button"
				aria-expanded={open}
				aria-controls={contentId}
				onClick={onHeaderClick}>
				{headerElement}
				<ChevronDown className={`Collapsible__arrow${open ? ' Collapsible__arrow--open' : ''}`} />
			</button>
			{cloneElement(contentElement, {
				id: contentId,
				className: addClassName(contentElement, 'Collapsible__body'),
				role: 'region',
				'aria-labelledby': headerId,
				ref: useCollapse(open),
			})}
		</div>
	);
};

Collapsible.propTypes = {
	/** The component id. */
	id: PropTypes.string,
	/** The component class name(s). */
	className: PropTypes.string,
	/** Whether the child component is displayed. */
	open: PropTypes.bool,
	/** The child components. */
	children: PropTypes.arrayOf(PropTypes.element).isRequired,
	/** Function invoked when the header is clicked. */
	onHeaderClick: PropTypes.func,
};

export default Collapsible;
