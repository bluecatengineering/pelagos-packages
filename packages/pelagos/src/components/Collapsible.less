@import '../../less/focus';
@import '../../less/fonts';
@import '../../less/spacing';
@import '../../less/utils';

@layer pelagos {
	.Collapsible {
		border-top: 1px solid var(--border-subtle);
	}

	.Collapsible__header {
		@focus-visible();
		@layer-hover-effects();
		display: flex;
		flex-direction: row;
		align-items: center;
		background-color: transparent;
		border: none;
		padding: @sp-04 @sp-16;
		cursor: pointer;
	}

	.Collapsible__title {
		@body-01();
		flex: 1;
		text-align: left;
	}

	.Collapsible__arrow {
		color: var(--link-primary);
		transition: transform 0.15s ease-out;

		&--open {
			transform: rotate(180deg);
		}

		@media print {
			display: none;
		}
	}

	.Collapsible__body {
		margin: @sp-08 @sp-16 @sp-24 @sp-16;
		overflow: hidden;
	}
}
