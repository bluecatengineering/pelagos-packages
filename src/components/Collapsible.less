@import '../../less/focus';
@import '../../less/fonts';
@import '../../less/spacing';

.Collapsible__arrowBox {
	color: var(--action-color);
	font-size: 21px;
	width: 16px;
	padding: @sp-02;
	margin-right: @sp-08;
	transition: background-color 0.15s ease-out;

	@media print {
		display: none;
	}
}

.Collapsible__arrow {
	transition: transform 0.15s ease-out;
}

.Collapsible__arrow--open {
	transform: rotate(90deg);
}

.Collapsible__header {
	@focus-visible();
	cursor: pointer;
	flex-direction: row;
	align-items: center;
	margin: @sp-02;

	&:hover > .Collapsible__arrowBox {
		background-color: var(--secondary-button-hover);
	}
	&:active > .Collapsible__arrowBox {
		background-color: var(--secondary-button-active);
	}
}

.Collapsible--l1 {
	@font-18-400();
}

.Collapsible--l2 {
	@font-16-600();
}

.Collapsible--l3 {
	@font-14-600();
}

.Collapsible--l4 {
	@font-12-600-uc();
	color: var(--font-color-2);
	margin-top: @sp-16;
}

.Collapsible--l5 {
	@font-10-600-uc();
	color: var(--font-color-2);
	margin-top: @sp-08;
}

.Collapsible__body {
	margin-bottom: @sp-16;
	overflow: hidden;
}
