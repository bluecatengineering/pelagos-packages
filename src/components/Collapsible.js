import {cloneElement, useMemo} from 'react';
import PropTypes from 'prop-types';
import {faAngleRight} from '@fortawesome/free-solid-svg-icons';

import handleButtonKeyDown from '../functions/handleButtonKeyDown';
import useCollapse from '../hooks/useCollapse';

import SvgIcon from './SvgIcon';
import './Collapsible.less';

/** A collapsible container. */
const Collapsible = ({id, className, open, header, level, children, onHeaderClick}) => {
	id = useMemo(() => id || 'e' + ('' + Math.random()).substr(2), [id]);
	const headerId = id + '-header';
	const contentId = id + '-content';
	return (
		<div id={id} className={`Collapsible${className ? ` ${className}` : ''}`}>
			<div
				id={headerId}
				className={`Collapsible__header Collapsible--l${level}`}
				tabIndex={0}
				role="button"
				aria-expanded={open}
				aria-controls={contentId}
				onClick={onHeaderClick}
				onKeyDown={handleButtonKeyDown}>
				<span className="Collapsible__arrowBox">
					<SvgIcon className={`Collapsible__arrow${open ? ' Collapsible__arrow--open' : ''}`} icon={faAngleRight} />
				</span>
				{header}
			</div>
			{cloneElement(children, {
				id: contentId,
				className: `${children.props.className ? children.props.className + ' ' : ''}Collapsible__body`,
				role: 'region',
				'aria-labelledby': headerId,
				ref: useCollapse(open),
			})}
		</div>
	);
};

Collapsible.propTypes = {
	/** The component id. */
	id: PropTypes.string,
	/** The component class name(s). */
	className: PropTypes.string,
	/** Whether the child component is displayed. */
	open: PropTypes.bool,
	/** The component header. */
	header: PropTypes.oneOfType([PropTypes.string, PropTypes.element]),
	/** The font size level. */
	level: PropTypes.oneOf(['1', '2', '3', '4', '5']),
	/** The child component. */
	children: PropTypes.element,
	/** Function invoked when the header is clicked. */
	onHeaderClick: PropTypes.func,
};
Collapsible.defaultProps = {
	level: '1',
};

export default Collapsible;
