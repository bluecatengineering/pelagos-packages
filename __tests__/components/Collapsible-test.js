import {shallow} from 'enzyme';

import Collapsible from '../../src/components/Collapsible';

jest.unmock('../../src/components/Collapsible');

describe('Collapsible', () => {
	describe('rendering', () => {
		it('renders expected elements when open is true and id is set', () => {
			const wrapper = shallow(
				<Collapsible id="test" open={true} header="Test">
					<div className="TestClass" />
				</Collapsible>
			);
			expect(wrapper.getElement()).toMatchSnapshot();
		});

		it('sets a random id if not provided', () => {
			const random = jest.spyOn(Math, 'random').mockReturnValue(0.1);
			const wrapper = shallow(
				<Collapsible open={false} header="Test">
					<div className="TestClass" />
				</Collapsible>
			);
			expect(wrapper.getElement()).toMatchSnapshot();
			expect(random.mock.calls).toEqual([[]]);
		});

		it('renders expected elements when className is set', () => {
			const wrapper = shallow(
				<Collapsible id="test" className="TestClass" open={true} header="Test">
					<div className="TestClass" />
				</Collapsible>
			);
			expect(wrapper.getElement()).toMatchSnapshot();
		});
	});

	describe('behaviour', () => {
		it('calls onHeaderClick when the header is clicked', () => {
			const onHeaderClick = jest.fn();
			const wrapper = shallow(
				<Collapsible id="test" header="Test" onHeaderClick={onHeaderClick}>
					<div />
				</Collapsible>
			);
			wrapper.find('#test-header').simulate('click');
			expect(onHeaderClick).toHaveBeenCalled();
		});
	});
});
