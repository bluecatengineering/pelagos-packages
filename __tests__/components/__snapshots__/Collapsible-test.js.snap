// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Collapsible rendering renders expected elements 1`] = `
<div
  className="Collapsible TestClass"
  id="test"
>
  <button
    aria-controls="test-content"
    aria-expanded={true}
    className="Collapsible__header"
    id="test-header"
    type="button"
  >
    cloneWithClassName
    <SvgIcon
      className="Collapsible__arrow Collapsible__arrow--open"
      icon={
        Object {
          "icon": Array [],
          "iconName": "angle-down",
          "prefix": "fas",
        }
      }
    />
  </button>
  <div
    aria-labelledby="test-header"
    className="addClassName"
    id="test-content"
    role="region"
  />
</div>
`;

exports[`Collapsible rendering renders expected elements when className is set 1`] = `
<div
  className="Collapsible TestClass"
  id="test"
>
  <button
    aria-controls="test-content"
    aria-expanded={true}
    className="Collapsible__header"
    id="test-header"
    type="button"
  >
    cloneWithClassName
    <SvgIcon
      className="Collapsible__arrow Collapsible__arrow--open"
      icon={
        Object {
          "icon": Array [],
          "iconName": "angle-down",
          "prefix": "fas",
        }
      }
    />
  </button>
  <div
    aria-labelledby="test-header"
    className="addClassName"
    id="test-content"
    role="region"
  />
</div>
`;

exports[`Collapsible rendering sets a random id if not provided 1`] = `
<div
  className="Collapsible"
  id="e1"
>
  <button
    aria-controls="e1-content"
    aria-expanded={false}
    className="Collapsible__header"
    id="e1-header"
    type="button"
  >
    cloneWithClassName
    <SvgIcon
      className="Collapsible__arrow"
      icon={
        Object {
          "icon": Array [],
          "iconName": "angle-down",
          "prefix": "fas",
        }
      }
    />
  </button>
  <div
    aria-labelledby="e1-header"
    className="addClassName"
    id="e1-content"
    role="region"
  />
</div>
`;
